Improve seq keysystem

krawthekrow opened the pull request:
Fixes #151, #152 and #153. ~For #152, I simply removed the line that checks if a keytab already exists -- is there a reason why it was there to begin with? If so, I could make it so that it checks for and ignores the error in seq.py instead. I can't quite think of a case when you wouldn't want to update the keytab if there's one already there.~ (Resolved in review)

celskeggs commented on building/build-debs/homeworld-admin-tools/src/setup.py:
```diff
@@ -129,7 +129,6 @@ def setup_keygateway(ops: Operations, config: configuration.Config) -> None:
         # keytab is stored encrypted in the configuration folder
         keytab = os.path.join(configuration.get_project(), "keytab.%s.crypt" % node.hostname)
         decrypted = keycrypt.gpg_decrypt_to_memory(keytab)
-        ops.ssh("confirm no existing keytab on @HOST", node, "test", "!", "-e", KEYTAB_PATH)
```
The original point here was to avoid overwriting a newer keytab with an older keytab.

We have a few options:

 * Your current change, which allows for the possibility of overwriting and losing a keytab.
 * Silently refuse to overwrite the keytab, which will cause pains when the keytab is updated, but will allow for re-deploying the same cluster.
 * Refuse to overwrite the keytab only if it has changed, which will cause pain (but not as much) when the keytab is updated, and allow for re-deploying the same cluster, but will be more complicated.
 * Refuse to overwrite the keytab only if it has changed and the updated keytab would have an older kvno than the last keytab, which will be the best solution, but will be much more complicated.

I would recommend the third solution, but I'm interested in any thoughts you have.

krawthekrow replied:
Would it be better to merge the keytabs? From what I understand, you would mainly be concerned about overwriting the existing keytab if there already exists a different keytab for other reasons. Maybe Spire can make a backup of the existing keytab and spit out a warning at the end of the session.

Or is the concern that the administrator may accidentally run the command with an outdated keytab? It doesn't sound like that would be a problem if the latest keytab is stored in, for example, Git.

ichung replied:
> ...the administrator may accidentally run the command with an outdated keytab? It doesn't sound like that would be a problem...

That sounds like precisely the sort of thing people will do accidentally.

krawthekrow replied:
I meant that if he does do that, he can just get the correct keytab from Git and replace it.

celskeggs replied:
Only if they realize it before their certificates expire. Otherwise, no one can regain access to the cluster until someone goes and drags the disaster recovery keys out of cold storage.

krawthekrow replied:
Right, it's encrypted with the disaster recovery key. Would merging the keytabs work here, then? It should be relatively simple to do it with ktutil; it's possible to scrape the kvnos from klist, but that feels ugly.

celskeggs replied:
Being encrypted with the disaster recovery key has nothing to do with it -- you'd need to pull the DR key out of storage simply because you have no other way to access the cluster.

Merging the keytabs might work, but I don't want that to silently occur. I think your idea of the third solution plus a force-overwrite option sounds good to me.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/verify.py:
```diff
@@ -75,7 +75,7 @@ def check_keygateway():
         access.call_keyreq("check")
     except subprocess.CalledProcessError as e:
-        command.fail("keygateway check failed: %s" % e)
+        command.fail("keygateway check failed (do you have valid Kerberos tickets?): %s" % e)
```
It might be worth detecting the exact failure mode ('returned non-zero exit status 1') and only providing the extra information in that case, and then also being louder about the the error, rather than sticking it in the middle of a longer error message.

Maybe an error like:

    keygateway check failed: [...]
    ** do you have valid kerberos tickets? **

celskeggs commented:
Sorry for not getting back to you sooner on this.

krawthekrow commented:
"Fixed" the other issue mentioned in the review; now fixes #160 too.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/command.py:
```diff
@@ -5,8 +5,11 @@ class CommandFailedException(Exception):
+    formatted_message = '\033[91m' + str(message)
+    if hint is not None:
+        formatted_message += '\n\033[93m' + str(hint)
+    raise CommandFailedException(formatted_message)
```
Besides the fact that this is being fixed elsewhere (note that #160 is claimed), it probably doesn't make sense to add color at this point: it probably makes more sense to add color when CommandFailedException is caught on the other end.

I do like your idea of having a "hint" section, however, so I recommend you expand CommandFailedException to have that as an additional attribute and print it out when CommandFailedException is caught.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/verify.py:
```diff
@@ -75,7 +75,9 @@ def check_keygateway():
-        command.fail("keygateway check failed: %s" % e)
+        command.fail("keygateway check failed: %s" % e,
+            '** do you have valid kerberos tickets? **' if
+            e.returncode == 1 else None)
```
I would avoid using an inline `if` here, for the sake of readability, but it doesn't matter that much.

krawthekrow commented:
Alright, everything has been responded to. I'll update it to use check_ssh after #171 gets finalized.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/setup.py:
```diff
@@ -122,18 +122,37 @@ def admit_keyserver(ops: Operations, config: configuration.Config) -> None:
+                    if existing_keytab != decrypted:
+                        command.fail("existing keytab does not match local keytab")
+                    print("existing keytab matches")
+                    return
```
You can probably encode this as a single ssh command, rather than needing to round-trip it multiple times and have all of this helper code involved.

krawthekrow replied:
Good idea, will implement.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/setup.py:
```diff
@@ -212,6 +231,7 @@ def wrap_param_tx(params):
     "keyserver": wrapop("deploy keys and configuration for keyserver; start keyserver", setup_keyserver),
     "self-admit": wrapop("admit the keyserver into the cluster during bootstrapping", admit_keyserver),
     "keygateway": wrapop("deploy keytab and start keygateway", setup_keygateway),
+    "update-keytab": wrapop("update keytab and restart keygateway", update_keytab),
```
These two commands should have clearer names. Also, keep in mind that the parameter to setup_keygateway will be treated as an optional command-line parameter, as far as I remember, which might not be what you want.

krawthekrow replied:
What names do you suggest? "Set up keygateway" and "update the keytab" sounds like exactly what those commands are for. Would it be better to actually have force-overwrite as an optional string parameter, so that the administrator would call "spire setup keygateway --force-overwrite" to update the keytab?

Also thanks for noting that, will fix.

celskeggs replied:
I'd name it `update-keygateway`, so that it's clear that the operations are very similar.

krawthekrow commented:
Not a trivial rebase. Moved upload_bytes to ssh.py, added type hints to functions in ssh.py, made hint yellow, and updated access_ssh to use hint system.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/access.py:
```diff
@@ -67,10 +67,8 @@ def access_ssh(add_to_agent=False):
-            print("or, perhaps your agent is broken; consider killing it and launching a new one.")
-            command.fail("*** ssh-add failed! ***")
-
+            fail_hint = "ssh-add returned non-zero exit code. do you have a ssh-agent?\nor, perhaps your agent is broken; consider killing it and launching a new one."
```
Might want to have this be multiple lines for readability, either with `"""..."""` notation or by having a second line with `fail_hint +=`.

celskeggs commented on building/build-debs/homeworld-admin-tools/src/command.py:
```diff
@@ -76,5 +78,9 @@ def main_invoke(command, params):
         print('\x1b[1;31m' + 'command failed: ' + str(e) + '\x1b[0m')
+        if e.hint is not None:
+            # yellow
+            print('\x1b[1;33m' + e.hint + '\x1b[0m')
```
We could make this clearer by putting our magic strings into named globals.

celskeggs commented on building/build-debs/homeworld-admin-tools/debian/changelog:
```diff
@@ -1,3 +1,9 @@
+homeworld-admin-tools (0.1.25); urgency=medium
```
This is missing the `stretch` keyword.

krawthekrow replied:
Oops, thanks!

krawthekrow commented:
Addressed issues. I also moved the error checking for check_keygateway into call_keyreq, since the same problems occur everywhere that call_keyreq is called, and those places do not appear to handle keyreq errors.
